zkllvm/use-cases/primer.mdx

```diff
@@ -0,0 +1,222 @@
+
+:::
+
+:::info
+
+This primer uses `crypto3` (C++) and `arkworks` (Rust) to show how to use external libraries in a circuit. The primer can be reused for any other suitable library.
+
+:::
+
+### Using CMake
+
+Switch to the working directory:
+
+```bash
+mkdir new_circuit && cd new_circuit
+```
+
+Create a directory for CMake modules:
+
+```bash
+mkdir cmake && cd cmake
+```
+
+Add the [**`CircuitCompile.cmake` module from the zkLLVM repository**](https://github.com/NilFoundation/zkLLVM/blob/master/cmake/CircuitCompile.cmake) to the `./new_circuit/cmake` directory:
+
+```
+new_circuit
+-- cmake
+    -- CircuitCompile.cmake
+```
+
+Create the circuit:
+
+```bash
+cd .. && mkdir src
+```
+
+```bash
+cd src && touch main.cpp
+```
+
+Add circuit code that references an external library:
+
+```cpp
+#include <nil/crypto3/hash/algorithm/hash.hpp>
+#include <nil/crypto3/hash/sha2.hpp>
+
+...
+
+[[circuit]] bool circuit_func() {...}
+```
+
+Create a configuration file for CMake:
+
+```bash
+cd .. && touch CMakeLists.txt
+```
+
+Inside `CMakeLists.txt`, set the project config and include the `CircuitCompile.cmake` module:
+
+```cmake
+cmake_minimum_required(VERSION 3.2)
+
+project(hashes_circuit
+    VERSION 1.0
+    DESCRIPTION "Tutorial circuit with hashes"
+    LANGUAGES CXX)
+
+list(APPEND CMAKE_MODULE_PATH "path/to/cmake")
+
+include(CircuitCompile)
+```
+
+Add the required library to the project configuration:
+
+```cmake
+set(crypto3_DIR "path/to/crypto3_headers")
+
+find_package(crypto3 REQUIRED)
+```
+
+:::tip[Boost]
+
+`crypto3` already depends on Boost. If an alternative library (that does not depend on Boost) is used, add Boost manually:
+
+```cmake
+find_package(Boost REQUIRED COMPONENTS system filesystem)
+```
+
+:::
+
+Set the circuit source and add a build target:
+
+```
+set(SOURCES src/main.cpp)
+
+add_circuit(
+    circuit SOURCES ${SOURCES}
+    LINK_LIBRARIES
+    crypto3::all
+    ${Boost_LIBRARIES}
+)
+
+set(binary_name circuit.ll)
+```
+
+, where `circuit` is the desired target name.
+
+Compile the circuit with:
+
+```bash
+cmake -G "Unix Makefiles" -B ./build -DCMAKE_BUILD_TYPE=Release .
+cd build
+make -C ./build circuit -j$(nproc)
+```
+
+:::warning
+
+The `CircuitCompile.cmake` module does not support using custom paths to `clang` built from sources (as well as custom paths to a linker). To use this module, it is necessary to [**install the `zkllvm` Deb package**](../getting-started/installation#install-zkllvm-binaries).
+
+:::
+
+:::tip
+
+Instead of specifying `crypto3_DIR` in `CMakeLists.txt`, it is possible to specify it when calling `cmake`:
+
+```bash
+cmake -G "Unix Makefiles" -B ./build -DCMAKE_BUILD_TYPE=Release . -Dcrypto3_DIR=/path/to/crypto3/installation
+```
+
+This method is useful if a dependency is installed in a custom location rather than `/usr/local/` or any other location requiring root permissions.
```

Comment:
Actually the permissions of a directory make no difference in specifying the path. I would say that this method could be useful in case if `crypto3_DIR` is not fixed (for example, a user is writing a script that installs circuit dependencies and then calls cmake), and one doesn't want to hardcode path in a file.

Reply:
Thank you, I have revised the explanation